chain/chunks/src/lib.rs

```diff
@@ -909,6 +906,39 @@ impl ShardsManager {
         Ok(header)
     }
 
+    pub fn insert_chunk_forward(&mut self, forward: PartialEncodedChunkForwardMsg) {
+        let chunk_hash = forward.chunk_hash.clone();
+        let num_total_parts = self.runtime_adapter.num_total_parts() as u64;
+        match self.chunk_forwards_cache.cache_get_mut(&chunk_hash) {
+            None => {
+                // Never seen this chunk hash before, collect the parts and cache them
+                let parts = forward.parts.into_iter().filter_map(|part| {
+                    let part_ord = part.part_ord;
+                    if part_ord > num_total_parts {
+                        warn!(target: "chunks", "Received chunk part with part_ord greater than the the total number of chunks");
+                        None
+                    } else {
+                        Some((part_ord, part))
+                    }
+                }).collect();
+                self.chunk_forwards_cache.cache_set(chunk_hash, parts);
+            }
+
+            Some(existing_parts) => {
+                for part in forward.parts {
+                    let part_ord = part.part_ord;
+                    if part_ord > num_total_parts {
+                        warn!(target: "chunks", "Received chunk part with part_ord greater than the the total number of chunks");
+                        continue;
+                    }
+                    if let Some(_) = existing_parts.insert(part_ord, part) {
+                        warn!(target: "chunks", "Part ord {} for chunk {:?} was forwarded multiple times", part_ord, chunk_hash.0);
```

Comment:
why would this be a concern? It seems to me that this would happen pretty regularly.

Reply:
I don't think it should happen. Nodes only forward the parts they own, and I assume each part has one owner, so each part should be forwarded exactly once.

Reply:
Let's say that the partial encoded chunk message to some block producer from the original chunk producer is lost due to a network glitch. In this case, couldn't it be that some other chunk producers already reconstruct the chunk and forward the same chunk parts to the same block producer while the block producer still doesn't have the chunk header?

Reply:
@bowenwang1996 you only forward parts you own. If you reconstruct the whole chunk, you do not send anything else optimistically

Reply:
I see. Makes sense.